File: mid/Greedy Algorithm/55. Jump Game.py
# Time complexity: O(n)
# Space complexity: O(1)

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        reach = 0
        length = len(nums) - 1

        for i in range(len(nums)):
            if i > reach:
                return False
            if reach >= length:
                return True
            reach = max(reach, i + nums[i])
        return reach >= length

File: mid/Greedy Algorithm/55. Jump Game.cpp
// Time complexity: O(n)
// Space complexity: O(1)

class Solution {
public:
    bool canJump(vector<int>& nums)
    {
        int length = nums.size();

        int reach = 0;
        for(int i = 0; i < length; i++)
        {
            if(i > reach)
                return false;
            if(reach > length)
                return true;
            reach = max(reach, nums[i] + i);
        }
        return true;
    }
};